crates/zk_evm/src/reference_impls/memory.rs

```diff
@@ -380,30 +380,35 @@ impl Memory for SimpleMemory {
 
     fn finish_global_frame(
         &mut self,
-        _base_page: MemoryPage,
+        base_page: MemoryPage,
         last_callstack_this: Address,
         returndata_fat_pointer: FatPointer,
         _timestamp: Timestamp,
     ) {
         // Safe to unwrap here, since `finish_global_frame` is never called with empty stack
-        let _current_observable_pages = self.observable_pages.current_frame();
-        let returndata_page = returndata_fat_pointer.memory_page;
+        let current_observable_pages = self.observable_pages.current_frame();
 
         // This is code oracle and some preimage has been decommitted into its memory.
         // We must keep this memory page forever for future decommits.
-        let is_returndata_page_static =
-            last_callstack_this == *CODE_ORACLE_ADDRESS && returndata_fat_pointer.length > 0;
-
-        // Note: we should not clean pages
-        // for &page in current_observable_pages {
-        // If the page's number is greater than or equal to the `base_page`,
-        // it means that it was created by the internal calls of this contract.
-        // We need to add this check as the calldata pointer is also part of the
-        // observable pages.
-        // if page >= base_page.0 && page != returndata_page {
-        //     self.memory.clear_page(page as usize);
-        // }
-        // }
+        let is_returndata_page_static = last_callstack_this == *CODE_ORACLE_ADDRESS;
+
+        // It is possible that the code oracle runs out of gas after decommitting code.
+        // In that case the page with the code is not the one that is returned.
+        let returndata_page = if is_returndata_page_static {
+            heap_page_from_base(base_page).0
+        } else {
+            returndata_fat_pointer.memory_page
```

Comment:
Btw in case of CodeOracle, it should return pointer to the heap already, isn't it?

Reply:
Only when it succeeds. If it panics, it returns a zero-size fat pointer to some garbage

Reply:
Interesting. So it won't be heap in this case?

Reply:
It is probably the 0th page on panic.